bundles/org.eclipse.e4.ui.css.swt/src/org/eclipse/e4/ui/css/swt/helpers/CSSSWTColorHelper.java

```diff
@@ -351,7 +352,13 @@ private static RGBA findColorByDefinition(String name) {
 	}
 
 	/** Helper function to avoid setting colors unnecessarily */
-	public static void setForeground(Control control, Color newColor) {
+	public static void setForeground(Control toolItem, Color newColor) {
+		if (!Objects.equals(toolItem.getForeground(), newColor)) {
```

Comment:
There is no need to guard against setting the same value, `ToolItem` implementations on all platforms already do this internally. Setting the same color repeatedly won't lead to unnecessary repaints.

Also, in SWT get and set for foreground/background aren't symmetric. get methods will delegate to the widget parent if the property isn't set, i.e. getForeground() actually means getEffectiveForeground(). The check above will incorrectly prevent setting the color if it matches the parent color. (Internal checks in SWT don't suffer from this.)

Reply:
Thanks, in this case I will move the calls to the corresponding CSS handlers.